File: backend/src/com/origin/entity/Grid.java
package com.origin.entity;

import org.jpark.TableExtended;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Index;
import javax.persistence.Table;
import java.sql.Blob;

/**
 * игровой "чанк" (регион), базовый кусок карты
 */
@Entity
@Table(name = "grids", indexes = {
		@Index(name = "id_uniq", columnList = "supergrid, x, y, level", unique = true),
		@Index(name = "coord", columnList = "x, y, level")
})
@TableExtended(creationSuffix = "engine=MyISAM")
public class Grid
{
	/**
	 * ид супергрида, по нему потом сделаем разбиение таблицы на партиции
	 * континент к которому оносится супергрид также зашит в ид супергрида
	 */
	@Column(name = "supergrid", columnDefinition = "INT(11) UNSIGNED NOT NULL", nullable = false)
	private int _supergrid;

	/**
	 * координаты грида внутри супергрида
	 */
	@Column(name = "x", columnDefinition = "INT(11) UNSIGNED NOT NULL")
	private int _x;

	@Column(name = "y", columnDefinition = "INT(11) UNSIGNED NOT NULL")
	private int _y;

	@Column(name = "level", columnDefinition = "INT(11) UNSIGNED NOT NULL")
	private int _level;

	/**
	 * сырые данные тайлов в виде массива байт, по 2 байта на 1 тайл
	 */
	@Column(name = "tiles", columnDefinition = "BLOB NOT NULL", nullable = false)
	private Blob _tilesBlob;

	public int getSupergrid()
	{
		return _supergrid;
	}

	public void setSupergrid(int supergrid)
	{
		_supergrid = supergrid;
	}

	public int getX()
	{
		return _x;
	}

	public void setX(int x)
	{
		_x = x;
	}

	public int getY()
	{
		return _y;
	}

	public void setY(int y)
	{
		_y = y;
	}

	public int getLevel()
	{
		return _level;
	}

	public void setLevel(int level)
	{
		_level = level;
	}

	public Blob getTilesBlob()
	{
		return _tilesBlob;
	}

	public void setTilesBlob(Blob tilesBlob)
	{
		_tilesBlob = tilesBlob;
	}
}

File: backend/src/com/origin/entity/Grid.kt
package com.origin.entity

import com.origin.model.Grid
import com.origin.model.LandLayer
import kotlinx.coroutines.ObsoleteCoroutinesApi
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.statements.api.ExposedBlob
import org.jetbrains.exposed.sql.transactions.transaction
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * игровой "чанк" (регион), базовый кусок карты
 * при больших объемах мира надо бить таблицу на партиции
 * по instance, суб партиции по x, y и тд
 */
object Grids : Table("grids") {
//    val id: Column<Int> = integer("id").autoIncrement()

    /**
     * на каком континенте находится грид, либо ид дома (инстанса, локации)
     */
    val region = integer("region")

    /**
     * координаты грида в мире (какой по счету грид, НЕ в игровых единицах)
     * разбиение таблицы (partitions) делаем на основе RANGE(x) и субпартициях на основе RANGE(y)
     */
    val x = integer("x")
    val y = integer("y")
    val level = integer("level")

    /**
     * время последнего обновления (в игровых тиках)
     */
    val lastTick = long("lastTick")

    /**
     * сырые данные тайлов в виде массива байт, по 2 байта на 1 тайл
     */
    val tilesBlob = blob("tiles")

//    override val primaryKey by lazy { super.primaryKey ?: PrimaryKey(id) }

    init {
        uniqueIndex(region, x, y, level)
    }

    override fun createStatement(): List<String> {
        return listOf(super.createStatement()[0] + " ENGINE=MyISAM")
    }
}

/**
 * НЕ DAO, потому что у нас хитрый индекс без явного id поля
 */
@ObsoleteCoroutinesApi
open class GridEntity(r: ResultRow, val layer: LandLayer) {
    //    val id = r[Grids.id]
    val region = r[Grids.region]
    val x = r[Grids.x]
    val y = r[Grids.y]
    val level = r[Grids.level]

    @Volatile
    var lastTick = r[Grids.lastTick]

    @Volatile
    var tilesBlob: ByteArray = r[Grids.tilesBlob].bytes

    companion object {
        /**
         * загрузка грида из базы
         */
        fun load(gx: Int, gy: Int, layer: LandLayer): Grid {
            val row = transaction {
                Grids.select { (Grids.x eq gx) and (Grids.y eq gy) and (Grids.level eq layer.level) and (Grids.region eq layer.region.id) }
                    .firstOrNull()
                    ?: throw RuntimeException("grid ($gx, $gy) level=${layer.level} region=${layer.region.id} is not found")
            }
            return Grid(row, layer)
        }

        val logger: Logger = LoggerFactory.getLogger(Grid::class.java)
    }

    /**
     * обновить blob с тайлами в базе
     */
    protected fun updateTiles() {
        Grids.update({ (Grids.x eq x) and (Grids.y eq y) and (Grids.region eq region) and (Grids.level eq level) }) {
            it[tilesBlob] = ExposedBlob(this@GridEntity.tilesBlob)
        }
    }
}
